src/utils/eip712/parser.rs

```diff
@@ -0,0 +1,144 @@
+use logos::Logos;
+
+use crate::utils::eip712::{error::Eip712Error, lexer::Token};
+use std::{fmt, result};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Type {
+    Address,
+    Uint,
+    Int,
+    String,
+    Bool,
+    Bytes,
+    Byte(u8),
+    Custom(String),
+    Array {
+        length: Option<u64>,
+        inner: Box<Type>,
+    },
+}
+
+impl From<Type> for String {
+    fn from(field_type: Type) -> String {
+        match field_type {
+            Type::Address => "address".into(),
+            Type::Uint => "uint".into(),
+            Type::Int => "int".into(),
+            Type::String => "string".into(),
+            Type::Bool => "bool".into(),
+            Type::Bytes => "bytes".into(),
+            Type::Byte(len) => format!("bytes{}", len),
+            Type::Custom(custom) => custom,
+            Type::Array { inner, length } => {
+                let inner: String = (*inner).into();
+                match length {
+                    None => format!("{}[]", inner),
+                    Some(length) => format!("{}[{}]", inner, length),
+                }
+            }
+        }
+    }
+}
+
+impl fmt::Display for Type {
+    fn fmt(&self, f: &mut fmt::Formatter) -> result::Result<(), fmt::Error> {
+        let item: String = self.clone().into();
```

Comment:
You clone the instance of `Type`, then create a string instance representing the display name of the type and create a string out of it. Basically throwing away both of the cloned instances that you shouldn't even need in the first place. If you just implement `as_str` for `Type` and call it when writing the staring, you would not need any of the clones.